### PR TITLE
Improve security schemes

### PR DESCRIPTION
**Pull request by grizio**
The current security schemes were not completely spec-compliant of Open API.
It answered a quick issue with a part of the spec only.

We take more time to be more spec compliant now.

This commit considers:

* https://swagger.io/specification/#securitySchemeObject
* https://swagger.io/docs/specification/authentication/

This commit does the following:

* Add parameters into security_scheme according to spec
* Rename `parameters` to `specification_extensions` to have the same name that in the spec
* Change the list of security schemes to be simpler
* Change the form of security schemes to consider different cases and attributes
* Add controls into model to ensure spec-compliance
* Correct swagger generation

# QA

<details>
  <summary>New list</summary>

image
</details>

<details>
  <summary>New form</summary>

image

image

image

image

</details>

<details>
  <summary>Corrected swagger</summary>

Copy/paste into https://editor.swagger.io/ to see the result.
:information_source: Responses in my project are not set.

```json
{
  "openapi": "3.0.0",
  "info": {
    "description": "URL\r\n** Please fill URL PROD if it exists or any other urls relative to this project\r\n\r\nFormat\r\n** Specify JSON / Restful format : https://i.imgur.com/rzoYLSX.png\r\n\r\nEncoding\r\n** Define encoding - utf-8 par défaut : https://i.imgur.com/j7pvEy1.png\r\n\r\nCompression\r\n** Define compression : https://i.imgur.com/Eix5nng.png\r\n\r\nProtocol\r\n** Make call with HTTPS (cf. Apple) : https://i.imgur.com/2a6mD2B.png\r\n\r\nLocalization\r\n** Define localization (header preferably or parameter) : https://i.imgur.com/NFDbaiL.png\r\n\r\nDates\r\n** Our date format should fullfill iso 8601 : https://fr.wikipedia.org/wiki/ISO_8601\r\n\r\nAuthentication\r\n** Define the format of the request and the authentication response =\u003e Example: authentication OAuth =\u003e Think to create a route!\r\n** Define errors and associated messages\r\n** Define the headers for each public / private route\r\n** Define the expiration logic of tokens\r\n",
    "title": "tst",
    "version": "1.0.0"
  },
  "servers": [],
  "tags": [
    {
      "name": "Report"
    }
  ],
  "paths": {
    "/reports": {
      "get": {
        "tags": [
          "Report"
        ]
      }
    },
    "/reports/{id}": {
      "get": {
        "tags": [
          "Report"
        ],
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          }
        ]
      }
    },
    "/reports/{id}/detail": {
      "get": {
        "tags": [
          "Report"
        ],
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          }
        ]
      }
    },
    "/reports/{id}/detail/{subid}": {
      "put": {
        "tags": [
          "Report"
        ],
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          },
          {
            "name": "subid",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          }
        ]
      }
    }
  },
  "components": {
    "schemas": {
      "DefaultReport": {
        "title": "Report - DefaultReport",
        "type": "object",
        "properties": {
          "id": {
            "type": "string"
          },
          "code": {
            "type": "string"
          }
        },
        "additionalProperties": false,
        "description": "Automatically generated (please edit me)",
        "required": [
          "code",
          "id"
        ]
      }
    },
    "securitySchemes": {
      "withOpenIdConnect": {
        "type": "openIdConnect",
        "description": "This is the description of the ~Security scheme~.",
        "openIdConnectUrl": "http://example.com"
      },
      "withOauth2": {
        "type": "oauth2",
        "description": "This is the description of the _Security scheme_.",
        "flows": {
          "authorizationCode": {
            "scopes": {
              "read": "Grants read access",
              "admin": "Grants access to admin operations",
              "write": "Grants write access"
            },
            "tokenUrl": "https://example.com/oauth/token",
            "authorizationUrl": "https://example.com/oauth/authorize"
          }
        }
      },
      "withHttp": {
        "type": "http",
        "description": "This is the description of the **Security scheme**.",
        "scheme": "basic"
      },
      "withApiKey": {
        "type": "apiKey",
        "description": "This is the description of the *Security scheme*.",
        "in": "header",
        "name": "Authorization"
      }
    }
  }
}
```
</details>

### CODE REVIEW

File: app/models/security_scheme.rb [hunk 1]
```diff
@@ -22,14 +41,49 @@ def parameters=(val)
...
-  def parameters_is_valid_json
-    JSON.parse(parameters) if parameters.is_a? String
+  def valid_type_option
+    errors.add(:security_scheme_type, 'invalid choice') if SecurityScheme.security_scheme_type_options[security_scheme_type].nil?
+  end
+
+  def valid_in_option
+    return if security_scheme_in.blank?
+    errors.add(:security_scheme_in, 'invalid choice') if SecurityScheme.security_scheme_in_options[security_scheme_in].nil?
+  end
```
Comment (grizio) [inline app/models/security_scheme.rb]: Workaround.
I failed to direct use the enum. When doing so, the value in ruby was always nil.

File: app/models/security_scheme.rb [hunk 2]
```diff
@@ -1,16 +1,35 @@
...
   validates :key, presence: true, uniqueness: { scope: :project }
   validates :security_scheme_type, presence: true
-  validates :name, presence: true
-  validates :security_scheme_in, presence: true
   validates :project, presence: true
 
-  validate :parameters_is_valid_json
-  validate :parameters_is_hash
+  validate :valid_type_option
+  validate :valid_in_option
```
Comment (Hugo-Hache) [inline app/models/security_scheme.rb]: You can use 
```
  validates :security_scheme_type, inclusion: { in: security_scheme_type_options.keys }
  validates :security_scheme_in, inclusion: { in: security_scheme_in_options.keys }, allow_blank: true
```

File: app/validators/json_validator.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class JsonValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    return if value.empty?
+    JSON.parse(value) if value.is_a? String
+  rescue JSON::ParserError
+    record.errors.add(attribute, value_must_be_valid_json) if value.is_a? String
```
Comment (Hugo-Hache) [inline app/validators/json_validator.rb]: I think a `:` is missing in front of `value_must_be_valid_json`

File: config/locales/security_schemes/en.yml [hunk 4]
```diff
@@ -2,3 +2,11 @@ en:
   security_schemes:
     index:
       no_security_scheme: "There are no security schemes, create one !"
+  activerecord:
+    errors:
+      models:
+        security_scheme:
+          attributes:
+            specification_extensions:
+              schema_must_be_valid_json: Must be a valid JSON object
```
Comment (Hugo-Hache) [inline config/locales/security_schemes/en.yml]: I think it should be `value_must_be_valid_json`

File: config/locales/security_schemes/en.yml [hunk 5]
```diff
@@ -2,3 +2,11 @@ en:
   security_schemes:
     index:
       no_security_scheme: "There are no security schemes, create one !"
+  activerecord:
```
Comment (Hugo-Hache) [inline config/locales/security_schemes/en.yml]: As your validator are now unrelated to a specific model, I think you should use this YAML hierarchy:

```
activerecord:
  errors:
    messages:
      value_must_be_valid_json: ...
      value_must_be_hash: ...
```

Source: https://guides.rubyonrails.org/i18n.html#error-message-scopes

File: app/validators/json_validator.rb (hunk 3 above)
Comment (grizio) [reply app/validators/json_validator.rb]: I renamed it but make a mistake. The next comment is a side-effect. :confounded:

File: config/locales/security_schemes/en.yml (hunk 5 above)
Comment (grizio) [reply config/locales/security_schemes/en.yml]: Hmm… I used what rails gave me. :thinking:

Comment (grizio) [reply config/locales/security_schemes/en.yml]: Ok, I understood

**Pull request merged**